src/dotnet/commands/dotnet-add-p2p/Program.cs

```diff
@@ -0,0 +1,232 @@
+                return ProjectRootElement.Open(filename);
+            }
+            catch (Microsoft.Build.Exceptions.InvalidProjectFileException)
+            {
+                return null;
+            }
+        }
+
+        public static ProjectRootElement GetProjectFromFileOrThrow(string filename)
+        {
+            if (!File.Exists(filename))
+            {
+                throw new GracefulException($"Provided project `{filename}` does not exist.");
+            }
+
+            var project = TryOpenProject(filename);
+            if (project == null)
+            {
+                throw new GracefulException($"Invalid MSBuild project `{filename}`.");
+            }
+
+            return project;
+        }
+
+        public static ProjectRootElement GetProjectFromCurrentDirectoryOrThrow()
+        {
+            DirectoryInfo currDir = new DirectoryInfo(Directory.GetCurrentDirectory());
+            FileInfo[] files = currDir.GetFiles("*proj");
+            if (files.Length == 0)
+            {
+                throw new GracefulException("Could not find any MSBuild project in the current directory.");
+            }
+
+            if (files.Length > 1)
+            {
+                throw new GracefulException("Found more than one MSBuild project in the current directory. Please specify which one to use.");
+            }
+
+            FileInfo projectFile = files.First();
+
+            if (!projectFile.Exists)
+            {
+                throw new GracefulException("Could not find any project in the current directory.");
+            }
+
+            var ret = TryOpenProject(projectFile.FullName);
+            if (ret == null)
+            {
+                throw new GracefulException($"Found an MSBuild project `{projectFile.FullName}` in the current directory but it is invalid.");
+            }
+
+            return ret;
+        }
+
+        public static Func<T, bool> AndPred<T>(params Func<T, bool>[] preds)
+        {
+            return (el) => preds.All((pred) => pred == null || pred(el));
+        }
+
+        public static string GetFrameworkConditionString(string framework)
+        {
+            if (string.IsNullOrEmpty(framework))
+            {
+                return null;
+            }
+
+            return $" '$(TargetFramework)' == '{framework}' ";
+        }
+
+        public static Func<T, bool> FrameworkPred<T>(string framework) where T : ProjectElement
+        {
+            string conditionStr = GetFrameworkConditionString(framework);
+            if (conditionStr == null)
+            {
+                return (ig) => {
+                    var condChain = ig.ConditionChain();
+                    return condChain.Count == 0;
+                };
+            }
+
+            conditionStr = conditionStr.Trim();
+            return (ig) => {
+                var condChain = ig.ConditionChain();
+                return condChain.Count == 1 && condChain.First().Trim() == conditionStr;
```

Comment:
We even do almost sorta kinda have an AST for conditions: https://github.com/Microsoft/msbuild/tree/xplat/src/XMakeBuildEngine/Evaluation/Conditionals. 

Filed https://github.com/Microsoft/msbuild/issues/1356 on your behalf, but I doubt we'll get to it soon, if ever.